domain/secret/service/watcher.go

```diff
@@ -187,8 +187,36 @@ func (s *WatchableService) WatchSecretsRotationChanges(ctx context.Context, owne
 	return newSecretWatcher(w, s.logger, processChanges)
 }
 
-func (s *WatchableService) WatchObsoleteUserSecrets(ctx context.Context) (watcher.NotifyWatcher, error) {
-	return watcher.TODO[struct{}](), nil
+// WatchObsoleteUserSecretsToPrune returns a watcher that notifies when a user secret revision is obsolete and ready to be pruned.
+func (s *WatchableService) WatchObsoleteUserSecretsToPrune(ctx context.Context) (watcher.NotifyWatcher, error) {
+	table, initialQuery := s.st.InitialWatchStatementForObsoleteUserSecretRevision()
+	wObsolete, err := s.watcherFactory.NewNamespaceWatcher(
+		table, changestream.Create, initialQuery,
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sWObsolete, err := newSecretWatcher(wObsolete, s.logger, s.st.GetObsoleteUserSecretRevisionsReadyToPrune)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	table, initialQuery = s.st.InitialWatchStatementForUserSecretsToPrune()
+	wAutoPrune, err := s.watcherFactory.NewNamespaceWatcher(
+		table, changestream.Update, initialQuery,
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sWAutoPrune, err := newSecretWatcher(wAutoPrune, s.logger, s.st.GetUserSecretRevisionsToPrune)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	w, err := eventsource.NewMultiStringsWatcher(ctx, sWObsolete, sWAutoPrune)
```

Comment:
Would it  be better to construct the underlying watchers as notification watchers, since we seem to be going to the effort to receive changesets and then just throwing away the data. If the underlying auto prune and obsolete watchers were notify, then this would be a MultiNotify watcher and we would not need the Normalise().
I guess the underlying changestream watcher is a StringsWatcher so our hands are tied. The notification watcher that is supported is a NewValueWatcher() but this is only for a specific entity.
We should discuss what we might want to do moving forward.

Reply:
Updated to use NamespaceNotifyWatcher introduced in https://github.com/juju/juju/pull/17592